qiskit/result/utils.py

```diff
@@ -118,10 +126,100 @@ def _adjust_creg_sizes(creg_sizes, indices):
     return new_creg_sizes
 
 
+def marginal_memory(
+    memory: List[str],
+    indices: Optional[List[int]] = None,
+    int_return: bool = False,
+    hex_return: bool = False,
+    parallel_threshold: int = 50,
+) -> Union[List[str], np.ndarray]:
+    """Marginalize memory
+
+    This function is multithreaded and will launch a thread pool with threads equal to the number
+    of CPUs by default. You can tune the number of threads with the ``RAYON_NUM_THREADS``
+    environment variable. For example, setting ``RAYON_NUM_THREADS=4`` would limit the thread pool
+    to 4 threads.
+
+    Args:
+        memory: The input memory list, this is a list of hexadecimal strings to be marginalized
```

Comment:
Does this not cover IQ data or is that outside the scope of this PR?

Reply:
I wasn't factoring that in when I wrote this, but we should try to support that in this function. I can add it to this PR if you have an example for what the input and output look like with IQ data.

Reply:
I added support for this in: https://github.com/Qiskit/qiskit-terra/pull/8051/commits/e709ce88e6acdd8f5031fd9989f93da25ed74f22 I still need to add testing to cover all the paths. But let me know if that interface works for you.

Reply:
fyi
https://github.com/Qiskit/qiskit-terra/blob/794db4792e97adca4a78eb4f3411c04e50665df7/qiskit/result/result.py#L211-L223

I really don't like how it is implemented without explicit typing.

Reply:
That table is what I based https://github.com/Qiskit/qiskit-terra/commit/e709ce88e6acdd8f5031fd9989f93da25ed74f22 but yeah the lack of explicit typing was annoying and why I needed an `avg_data` kwarg to differentiate between single level 1 and avg level 0 because I couldn't figure out a way to reliably detect the difference without an explicit input type. I was trying to make this function work independently of the `Results` object which is the only place I think that metadata would be stored.